docs/typecasting.rst

```diff
@@ -0,0 +1,175 @@
+you are introducing your own types, you will need to make sure they
+can be saved and loaded correctly.
+
+Some formats such as `date`, `time` and `datetime` may have
+additional options to it::
+
+    $m->addField('registered', [
+        'type'=>'date',
+        'persist_format'=>'d/m/Y',
+        'persist_timezone'=>'IST'
+    ]);
+
+Here is another example with booleans::
+
+    $m->addField('is_married', [
+        'type' => boolean',
+        'enum' => ['No', 'Yes']
+    ]);
+
+    $m['is_married'] = 'Yes';  // type-casts into true
+    $m['is_married'] = true;   // better way!
+
+    $m->save();   // stores as "Yes".
+
+Value types
+===========
+
+Any type can have a value of `null`::
+
+    $m['is_married'] = null;
+    if (!$m['is_married']) {
+        // either null or false
+    }
+
+If type is passed with incompatible format, Agile Data will try
+to cast the data::
+
+    $m->addField('age', ['type'=>'integer']);
+    $m->addField('name', ['type'=>'string']);
+
+    $m['age'] = '49.80';
+    $m['name'] = '       John';
+
+    echo $m['age']; // 49
+    echo $m['name']; // 'John'
+
+Undefined type
+--------------
+If you do not set type for a field, Agile Data will attempt to
+keep value unmodified.
+
+Because of the loose PHP types, you can encounter situations
+where undefined type is changed from `'4'` to `4`. This change
+is still considered "dirty".
+
+If you use with a typeless numeric value in SQL, the response
+from SQL does not distinguish between integers and strings, so
+your value will be stored as "string" inside the model.
+
+The same can be said about forms, which submit all their data
+through POST request that has no types, so undefined type 
+fields should work relatively good with the standard setup
+of Agile Data + Agile Toolkit + SQL.
+
+Type of IDs
+-----------
+
+Many databases will allow you to use different types for ID fields.
+In SQL the 'id' column will usually be "integer" but sometimes it
+can be a different value.
+
+The same applies for references ($m->hasOne()). 
+
+Supported types
+---------------
+
+- 'string' - for storing short strings, such as name of a person.
+  Normalize will trim the value.
+- 'integer' - normalize will (int) the value. 
+- 'array' - 
+
+Types and UI
+------------
+
+UI framework such as Agile Toolkit will typically rely on type
+information to properly present data with views (forms and tables)
+without you having to specify the `ui` property.
+
+Serialization
+=============
+
+Some types cannot be stored natively. For example, generic objects
+and arrays have no native type in SQL database. This is where
+serialization feature is used.
+
+Field may use serialization to further encode field value for the
+storage purpose::
+
+    $this->addField('private_key', [
+        'serialize'=>'base64',
+        'system'=>true,
+    ]);
+
+This is one way to store binary data. Type is unspecified but the
+binary value of a field will be encoded with base64 before storing.
+
```

Comment:
Currently AD supports following serialization algorithms:
* serialize - for storing PHP objects
* json - for storing objects and arrays
* base64 - for storing encoded strings
* [serialize_callback, unserialize_callback] - for custom serialization